File: lbm_cylinder.py
import matplotlib.pyplot as plt
import numpy as np
import os

"""
Python Program: Lattice Boltzmann Method (D2Q9) - Fluid Flow Past a Fixed Cylinder
================================================================================

This program simulates 2D fluid flow past a circular cylinder using the Lattice Boltzmann Method
with a D2Q9 lattice model. The simulation computes the evolution of fluid particle distribution
functions and visualizes the resulting flow field including vorticity and streamlines.

Key Features:
- D2Q9 lattice model with 9 velocity directions
- Reynolds number based flow simulation (Re = 200)
- Cylinder boundary condition handling with bounce-back method
- Periodic boundary conditions for domain edges
- Real-time visualization of vorticity and streamlines
- Organized output to dedicated 'python' subdirectory
- Optimized simulation parameters for faster execution
- Performance mode: disable visualization for computational benchmarking

What This Code Does:
====================
1. Sets up a 400x100 computational domain with a cylinder at position (70, 50)
2. Implements the LBM algorithm with streaming and collision steps
3. Applies proper boundary conditions (no-slip at cylinder, periodic at edges)
4. Computes macroscopic variables (density, velocity) from distribution functions
5. Visualizes the flow field showing vorticity patterns and streamlines (optional)
6. Saves output images every 2000 steps for analysis (optional)

Physics Background:
==================
The Lattice Boltzmann Method is a computational fluid dynamics technique that:
- Models fluid as discrete particles moving on a regular lattice
- Uses distribution functions to represent particle populations
- Replaces the Navier-Stokes equations with simple collision and streaming rules
- Naturally handles complex boundaries and multiphase flows
- Provides accurate results for incompressible flows at moderate Reynolds numbers

This simulation specifically studies the classic problem of flow past a cylinder,
which exhibits phenomena like boundary layer separation, vortex shedding, and
wake formation - fundamental concepts in fluid dynamics.

Author: Bart Blockmans
Date: August 2025
"""

# Directories
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
OUT_DIR = os.path.join(BASE_DIR, "python")
os.makedirs(OUT_DIR, exist_ok=True)

def initialize_parameters():
    """
    Initialize all simulation parameters including domain size, lattice properties,
    cylinder geometry, and fluid properties.
    """
    # Fluid domain dimensions
    MAX_X = 400
    MAX_Y = 100
    
    # D2Q9 lattice model parameters
    # 9 velocity directions: center (0), cardinal directions (1-4), diagonal directions (5-8)
    LATTICE_NUM = 9
    CX = np.array([0, 1, 0, -1, 0, 1, -1, -1, 1])    # x-velocity components
    CY = np.array([0, 0, 1, 0, -1, 1, 1, -1, -1])    # y-velocity components
    WEIGHTS = np.array([4/9, 1/9, 1/9, 1/9, 1/9, 1/36, 1/36, 1/36, 1/36])  # lattice weights
    OPP = np.array([0, 3, 4, 1, 2, 7, 8, 5, 6])      # opposite direction indices
    
    # Cylinder geometry and position
    POSITION_OX = 70    # x-coordinate of cylinder center
    POSITION_OY = 50    # y-coordinate of cylinder center
    RADIUS = 20         # cylinder radius
    
    # Fluid properties and Reynolds number
    REYNOLDS = 200      # Reynolds number for the flow
    U_MAX = 0.1        # maximum inlet velocity
    kinematic_viscosity = U_MAX * 2 * RADIUS / REYNOLDS
    relaxation_time = 3.0 * kinematic_viscosity + 0.5
    
    # Simulation control parameters
    MAX_STEP = 20001    # total number of time steps
    OUTPUT_STEP = 2000  # frequency of output visualization
    PICTURE_NUM = 1     # counter for saved images
    
    # Visualization control flags
    # Set to False for performance benchmarking (no visualization or file I/O)
    # Set to True for normal operation with visualization and output
    VISUALIZE = True
    
    # Clean visualization parameter: removes ticks, labels, title and minimizes margins
    # Set to True for clean, publication-ready images without visual clutter
    NOTICKS = False
    
    return (MAX_X, MAX_Y, LATTICE_NUM, CX, CY, WEIGHTS, OPP, 
            POSITION_OX, POSITION_OY, RADIUS, REYNOLDS, U_MAX, 
            kinematic_viscosity, relaxation_time, MAX_STEP, OUTPUT_STEP, PICTURE_NUM, VISUALIZE, NOTICKS)

def create_cylinder_mask(MAX_X, MAX_Y, POSITION_OX, POSITION_OY, RADIUS):
    """
    Create a boolean mask representing the cylinder in the computational domain.
    """
    x, y = np.meshgrid(range(MAX_X), range(MAX_Y))
    cylinder = (x - POSITION_OX)**2 + (y - POSITION_OY)**2 <= RADIUS**2
    return cylinder

def initialize_flow_field(MAX_Y, MAX_X, LATTICE_NUM, U_MAX, CX, CY, WEIGHTS):
    """
    Initialize the flow field with uniform density and inlet velocity.
    Sets up the initial distribution functions F based on equilibrium conditions.
    """
    # Initialize density and velocity fields
    rho = np.ones([MAX_Y, MAX_X])           # uniform density field
    ux, uy = np.zeros([MAX_Y, MAX_X]), np.zeros([MAX_Y, MAX_X])  # zero velocity initially
    ux[:, 0], ux[:, -1] = U_MAX, U_MAX     # set inlet/outlet velocity
    
    # Initialize distribution functions F using equilibrium distribution
    F = np.zeros([MAX_Y, MAX_X, LATTICE_NUM])
    for i, cx, cy, w in zip(range(LATTICE_NUM), CX, CY, WEIGHTS):
        F[:, :, i] = rho * w * (1 + 3 * (cx * ux + cy * uy) 
                                + 9 * (cx * ux + cy * uy)**2 / 2 
                                - 3 * (ux**2 + uy**2) / 2)
    return rho, ux, uy, F

def apply_periodic_boundary_conditions(F, MAX_Y, MAX_X):
    """
    Apply periodic boundary conditions to the distribution functions.
    This ensures particles leaving one edge of the domain re-enter from the opposite edge.
    """
    # Periodic boundary in x-direction (left/right edges)
    F[:, 0, [1, 5, 8]] = F[:, -1, [1, 5, 8]]    # particles moving right at left edge
    F[:, -1, [3, 6, 7]] = F[:, 0, [3, 6, 7]]    # particles moving left at right edge
    
    # Periodic boundary in y-direction (top/bottom edges)
    F[0, :, [2, 5, 6]] = F[-1, :, [2, 5, 6]]    # particles moving up at bottom edge
    F[-1, :, [4, 7, 8]] = F[0, :, [4, 7, 8]]    # particles moving down at top edge

def streaming_step(F, LATTICE_NUM, CX, CY):
    """
    Perform the streaming step of the LBM algorithm.
    Particles move along their velocity directions to neighboring lattice sites.
    """
    for i, cx, cy in zip(range(LATTICE_NUM), CX, CY):
        F[:, :, i] = np.roll(F[:, :, i], cx, axis=1)  # stream in x-direction
        F[:, :, i] = np.roll(F[:, :, i], cy, axis=0)  # stream in y-direction

def handle_cylinder_boundary(F, cylinder, LATTICE_NUM, CX, CY, OPP):
    """
    Handle the no-slip boundary condition at the cylinder surface.
    Implements the bounce-back method where particles hitting the cylinder
    reverse their direction.
    """
    cylinderF = F.copy()
    for i in range(1, LATTICE_NUM):  # skip the rest particle (i=0)
        # Find incoming particles: inside cylinder but moving toward cylinder surface
        incoming_particles = (cylinder & 
                            (np.roll(cylinder, -CX[i], axis=1) & 
                             np.roll(cylinder, -CY[i], axis=0) == False))
        # Bounce back: reverse particle direction
        cylinderF[incoming_particles, i] = F[incoming_particles, OPP[i]]
    return cylinderF

def compute_macroscopic_variables(F, cylinder, LATTICE_NUM, CX, CY):
    """
    Compute macroscopic fluid variables (density and velocity) from the
    distribution functions using moment integrals.
    """
    # Density: sum of all distribution functions
    rho = np.sum(F, axis=2)
    
    # Velocity: momentum divided by density
    ux = np.sum(F * CX, axis=2) / rho
    uy = np.sum(F * CY, axis=2) / rho
    
    # Set velocity to zero inside the cylinder (no-slip condition)
    ux[cylinder] = 0
    uy[cylinder] = 0
    
    return rho, ux, uy

def collision_step(F, rho, ux, uy, LATTICE_NUM, CX, CY, WEIGHTS, relaxation_time):
    """
    Perform the collision step using the BGK (Bhatnagar-Gross-Krook) approximation.
    Distribution functions relax toward their equilibrium values.
    """
    # Compute equilibrium distribution functions
    Feq = np.zeros(F.shape)
    for i, cx, cy, w in zip(range(LATTICE_NUM), CX, CY, WEIGHTS):
        Feq[:, :, i] = rho * w * (1 + 3 * (cx * ux + cy * uy) 
                                  + 9 * (cx * ux + cy * uy)**2 / 2 
                                  - 3 * (ux**2 + uy**2) / 2)
    
    # BGK collision: relax toward equilibrium
    F += -(1 / relaxation_time) * (F - Feq)

def apply_inflow_outflow_boundary_conditions(F, rho, ux, LATTICE_NUM, CX, WEIGHTS, U_MAX):
    """
    Apply velocity boundary conditions at inlet and outlet.
    Maintains constant velocity at domain boundaries.
    """
    # Set boundary velocities
    ux[:, 0], ux[:, -1] = U_MAX, U_MAX
    
    # Update distribution functions at boundaries using equilibrium distribution
    for i, cx, w in zip(range(LATTICE_NUM), CX, WEIGHTS):
        # Inlet boundary (left edge)
        F[:, 0, i] = rho[:, 0] * w * (1 + 3 * (cx * ux[:, 0]) 
                                      + 9 * (cx * ux[:, 0])**2 / 2 
                                      - 3 * (ux[:, 0]**2) / 2)
        # Outlet boundary (right edge)
        F[:, -1, i] = rho[:, -1] * w * (1 + 3 * (cx * ux[:, -1]) 
                                        + 9 * (cx * ux[:, -1])**2 / 2 
                                        - 3 * (ux[:, -1]**2) / 2)

def visualize_flow_field(step, ux, uy, cylinder, POSITION_OX, POSITION_OY, RADIUS, 
                        MAX_Y, MAX_X, PICTURE_NUM, NOTICKS):
    """
    Visualize the flow field by plotting vorticity and streamlines.
    Saves the plot as an image file.
    
    When NOTICKS is True, creates clean images without ticks, labels, title,
    and with minimized margins for publication use.
    """
    # Compute vorticity (curl of velocity field)
    vorticity = ((np.roll(uy, -1, axis=1) - np.roll(uy, 1, axis=1)) 
                 - (np.roll(ux, -1, axis=0) - np.roll(ux, 1, axis=0)))
    vorticity[cylinder] = np.nan  # hide vorticity inside cylinder
    
    # Create the visualization
    plt.imshow(vorticity, cmap="RdBu", origin="lower", vmin=-0.02, vmax=0.02)
    
    # Add cylinder outline
    plt.gca().add_patch(plt.Circle((POSITION_OX, POSITION_OY), RADIUS, color="black"))
    
    # Add streamlines colored by velocity magnitude
    Y, X = np.mgrid[0:MAX_Y, 0:MAX_X]
    speed = np.sqrt(ux**2 + uy**2)
    plt.streamplot(X, Y, ux, uy, color=speed, linewidth=1, cmap='cool')
    
    # Apply NOTICKS styling if requested
    if NOTICKS:
        # Clean layout: no ticks, labels, title, minimized margins
        plt.xticks([])
        plt.yticks([])
        plt.xlabel('')
        plt.ylabel('')
        plt.title('')
        # Minimize margins while keeping the plot box
        plt.subplots_adjust(left=0.05, right=0.95, top=0.95, bottom=0.05)
    else:
        # Normal layout: standard title
        plt.title(f'Vorticity and streamlines at step {step}')
    
    # Figure limits
    plt.xlim([0, MAX_X])
    plt.ylim([0, MAX_Y])

    # Set title and save
    plt.savefig(os.path.join(OUT_DIR, f"lattice_boltzmann_{PICTURE_NUM:04d}_python.png"))
    plt.pause(0.01)  # brief pause for real-time visualization
    plt.cla()        # clear current axes for next iteration
    
    return PICTURE_NUM + 1

def main():
    """
    Main simulation loop implementing the Lattice Boltzmann Method.
    
    This function orchestrates the complete LBM simulation by:
1. Setting up the computational domain and initial conditions
2. Running the main time-stepping loop with 7 key algorithm steps
3. Generating visualizations at regular intervals (optional)
4. Saving results to the 'python' output directory (optional)
    """
    # Initialize all parameters
    (MAX_X, MAX_Y, LATTICE_NUM, CX, CY, WEIGHTS, OPP, 
     POSITION_OX, POSITION_OY, RADIUS, REYNOLDS, U_MAX, 
     kinematic_viscosity, relaxation_time, MAX_STEP, OUTPUT_STEP, PICTURE_NUM, VISUALIZE, NOTICKS) = initialize_parameters()
    
    # Print simulation parameters for user reference
    print(f"Reynolds number = {REYNOLDS}")
    print(f"Relaxation time = {relaxation_time}")
    print(f"Domain size: {MAX_X} x {MAX_Y}")
    print(f"Maximum velocity = {U_MAX}")
    print(f"Kinematic viscosity = {kinematic_viscosity}")
    print(f"Visualization enabled: {VISUALIZE}")
    if VISUALIZE:
        print(f"Output directory: {OUT_DIR}")
        print(f"Clean visualization (NOTICKS): {NOTICKS}")
    
    # Create cylinder mask (defines where the solid cylinder is located)
    cylinder = create_cylinder_mask(MAX_X, MAX_Y, POSITION_OX, POSITION_OY, RADIUS)
    
    # Initialize flow field with uniform density and inlet velocity
    rho, ux, uy, F = initialize_flow_field(MAX_Y, MAX_X, LATTICE_NUM, U_MAX, CX, CY, WEIGHTS)
    
    # Main simulation loop - each iteration represents one time step
    print(f"Starting simulation for {MAX_STEP} steps...")
    for step in range(MAX_STEP):
        if step % 1000 == 0:  # Progress indicator every 1000 steps
            print(f"Step {step}/{MAX_STEP}")
        
        # LBM Algorithm Steps (executed in sequence each time step):
        
        # 1. Apply periodic boundary conditions
        #    Particles leaving one edge re-enter from the opposite edge
        apply_periodic_boundary_conditions(F, MAX_Y, MAX_X)
        
        # 2. Streaming step: particles move along their velocity directions
        #    This is the advection part of the LBM
        streaming_step(F, LATTICE_NUM, CX, CY)
        
        # 3. Handle cylinder boundary (no-slip condition)
        #    Particles hitting the cylinder bounce back (bounce-back method)
        cylinderF = handle_cylinder_boundary(F, cylinder, LATTICE_NUM, CX, CY, OPP)
        
        # 4. Compute macroscopic variables (density, velocity)
        #    These are the physical quantities we actually care about
        rho, ux, uy = compute_macroscopic_variables(cylinderF, cylinder, LATTICE_NUM, CX, CY)
        F = cylinderF
        
        # 5. Collision step: relaxation toward equilibrium
        #    This is the collision part of the LBM (BGK approximation)
        collision_step(F, rho, ux, uy, LATTICE_NUM, CX, CY, WEIGHTS, relaxation_time)
        
        # 6. Apply inlet/outlet boundary conditions
        #    Maintains constant velocity at domain boundaries
        apply_inflow_outflow_boundary_conditions(F, rho, ux, LATTICE_NUM, CX, WEIGHTS, U_MAX)
        
        # 7. Visualization and output (every OUTPUT_STEP iterations)
        #    Generates plots showing vorticity and streamlines
        if VISUALIZE and step % OUTPUT_STEP == 0:
            PICTURE_NUM = visualize_flow_field(step, ux, uy, cylinder, POSITION_OX, 
                                             POSITION_OY, RADIUS, MAX_Y, MAX_X, PICTURE_NUM, NOTICKS)
    
    print("Simulation completed successfully!")
    if VISUALIZE:
        print(f"Output images saved in: {OUT_DIR}")
    else:
        print("Performance mode: no visualization or file I/O performed.")

if __name__ == "__main__":
    main()

File: fortran_lbm/lbm_cylinder.f90
! LBM with half-way bounce-back (matching Python implementation exactly)
program lbm_halfway_bounceback
  implicit none
  
  ! Parameters
  integer, parameter :: NX = 400, NY = 100, Q = 9
  integer, parameter :: CX0 = 70, CY0 = 50, R = 20
  real, parameter :: UMAX = 0.1, REYNOLDS = 200.0
  real :: TAU  ! Relaxation time (calculated dynamically)
  integer :: NSTEPS, OUTPUT_STEP
  logical :: GENERATE_IMAGES
  
  ! Arrays
  real :: F(NY,NX,Q), F2(NY,NX,Q)
  real :: RHO(NY,NX), UX(NY,NX), UY(NY,NX)
  logical :: CYL(NY,NX)
  ! INCOMING_MASKS array removed - using original bounce-back approach
  real :: TMP2D(NY,NX)  ! 2D scratch buffer for streaming
  
  ! Lattice vectors
  integer :: CX(Q) = [0, 1, 0, -1, 0, 1, -1, -1, 1]
  integer :: CY(Q) = [0, 0, 1, 0, -1, 1, 1, -1, -1]
  integer :: OPP(Q) = [1, 4, 5, 2, 3, 8, 9, 6, 7]  ! Opposite directions (1-based)
  real :: W(Q) = [4.0/9.0, 1.0/9.0, 1.0/9.0, 1.0/9.0, 1.0/9.0, &
                  1.0/36.0, 1.0/36.0, 1.0/36.0, 1.0/36.0]
  
  integer :: i, j, k, step, pic
  real :: feq, vort(NY,NX)
  real :: cu(NY,NX), usq(NY,NX)
  real :: usqL(NY), usqR(NY)
  
  ! Parse command line arguments
  call parse_command_line_arguments()
  
  ! Calculate relaxation time dynamically (matching Python)
  TAU = 3.0 * (UMAX * 2.0 * real(R) / REYNOLDS) + 0.5
  
  write(*,*) 'LBM with Half-way Bounce-back (Python-matching)'
  write(*,*) 'Reynolds number =', REYNOLDS
  write(*,*) 'Relaxation time =', TAU
  write(*,*) 'Domain size =', NX, 'x', NY
  write(*,*) 'Umax =', UMAX
  write(*,*) 'NSTEPS =', NSTEPS
  write(*,*) 'OUTPUT_STEP =', OUTPUT_STEP
  write(*,*) 'Generate images =', GENERATE_IMAGES
  
  ! Initialize cylinder
  do j = 1, NX
    do i = 1, NY
      CYL(i,j) = ((j-CX0)**2 + (i-CY0)**2 <= R*R)
    end do
  end do
  write(*,*) 'Cylinder center: (', CX0, ',', CY0, '), radius:', R
  write(*,*) 'Cylinder mask:', count(CYL), 'solid cells out of', NY*NX
  
  ! No need for precomputed masks with original bounce-back approach
  
  ! Initialize flow
  RHO = 1.0
  UX = 0.0
  UY = 0.0
  UX(:,1) = UMAX
  UX(:,NX) = UMAX
  
  ! Initialize distribution functions
  do k = 1, Q
    do j = 1, NX
      do i = 1, NY
        cu(i,j) = real(CX(k))*UX(i,j) + real(CY(k))*UY(i,j)
        usq(i,j) = UX(i,j)**2 + UY(i,j)**2
        feq = RHO(i,j) * W(k) * (1.0 + 3.0*cu(i,j) + 4.5*cu(i,j)*cu(i,j) - 1.5*usq(i,j))
        F(i,j,k) = feq
      end do
    end do
  end do
  
  write(*,*) 'Starting simulation for', NSTEPS, 'steps...'
  pic = 1
  
  ! Main loop - matching Python algorithm exactly
  do step = 0, NSTEPS-1
    if (mod(step, 1000) == 0) then
      write(*,*) 'Step', step, '/', NSTEPS
      write(*,*) '  UX range:', minval(UX), 'to', maxval(UX)
      write(*,*) '  UY range:', minval(UY), 'to', maxval(UY)
      write(*,*) '  RHO range:', minval(RHO), 'to', maxval(RHO)
    end if
    
    ! Step 1: Apply periodic boundary conditions
    call apply_periodic_boundary_conditions()
    
    ! Step 2: Streaming step: particles move along their velocity directions
    call streaming_step()
    
    ! Step 3: Handle cylinder boundary (no-slip condition)
    call handle_cylinder_boundary()
    
    ! Step 4: Compute macroscopic variables
    call compute_macroscopic_variables()
    
    ! Step 5: Collision step: relaxation toward equilibrium
    call collision_step()
    
    ! Step 6: Apply inlet/outlet boundary conditions
    call apply_inflow_outflow_boundary_conditions()
    
    ! Step 7: Visualization
    if (GENERATE_IMAGES .and. mod(step, OUTPUT_STEP) == 0) then
      call plot_vorticity_png(step, pic)
      pic = pic + 1
    end if
  end do
  
  write(*,*) 'Simulation completed successfully!'
  if (GENERATE_IMAGES) then
    write(*,*) 'Images saved in ./images'
  else
    write(*,*) 'No images generated (GENERATE_IMAGES = .false.)'
  end if
  
contains

  subroutine parse_command_line_arguments()
    ! Parse command line arguments for NSTEPS, OUTPUT_STEP, and GENERATE_IMAGES
    integer :: argc, i, img_flag
    character(len=256) :: arg
    
    ! Set default values
    NSTEPS = 20001
    OUTPUT_STEP = 2000
    GENERATE_IMAGES = .true.
    
    ! Get number of command line arguments
    argc = command_argument_count()
    
    ! Check for help
    if (argc > 0) then
      call get_command_argument(1, arg)
      if (arg(1:2) == '-h' .or. arg(1:2) == '--') then
        write(*,*) 'Usage: ./lbm_cylinder_opt [NSTEPS] [OUTPUT_STEP] [GENERATE_IMAGES]'
        write(*,*) '  NSTEPS: Number of simulation steps (default: 20001)'
        write(*,*) '  OUTPUT_STEP: Steps between image outputs (default: 2000)'
        write(*,*) '  GENERATE_IMAGES: 1 to generate images, 0 to skip (default: 1)'
        write(*,*) ''
        write(*,*) 'Examples:'
        write(*,*) '  ./lbm_cylinder_opt                    ! Use defaults'
        write(*,*) '  ./lbm_cylinder_opt 10000 1000 1      ! 10k steps, output every 1k steps'
        write(*,*) '  ./lbm_cylinder_opt 5000 500 0        ! 5k steps, no images'
        stop
      end if
    end if
    
    ! Parse arguments
    do i = 1, argc
      call get_command_argument(i, arg)
      
      if (i == 1) then
        ! First argument: NSTEPS
        read(arg, *) NSTEPS
      else if (i == 2) then
        ! Second argument: OUTPUT_STEP
        read(arg, *) OUTPUT_STEP
      else if (i == 3) then
        ! Third argument: GENERATE_IMAGES (0 or 1)
        read(arg, *) img_flag
        GENERATE_IMAGES = (img_flag == 1)
      end if
    end do
    
    ! Validate arguments
    if (NSTEPS < 1) then
      write(*,*) 'Error: NSTEPS must be >= 1'
      stop
    end if
    
    if (OUTPUT_STEP < 1) then
      write(*,*) 'Error: OUTPUT_STEP must be >= 1'
      stop
    end if
    
    if (OUTPUT_STEP > NSTEPS) then
      write(*,*) 'Warning: OUTPUT_STEP > NSTEPS, no images will be generated'
    end if
  end subroutine parse_command_line_arguments

  ! Precomputed masks function removed - using original bounce-back approach
  
  ! roll_logical function removed - no longer needed with original bounce-back approach
  
  subroutine apply_periodic_boundary_conditions()
    ! Apply periodic boundary conditions (matching Python)
    do i = 1, NY
      F(i,1,2) = F(i,NX,2)    ! east at left edge
      F(i,1,6) = F(i,NX,6)    ! ne at left edge
      F(i,1,9) = F(i,NX,9)    ! se at left edge
      F(i,NX,4) = F(i,1,4)    ! west at right edge
      F(i,NX,7) = F(i,1,7)    ! nw at right edge
      F(i,NX,8) = F(i,1,8)    ! sw at right edge
    end do
    
    do j = 1, NX
      F(1,j,3) = F(NY,j,3)    ! north at bottom edge
      F(1,j,6) = F(NY,j,6)    ! ne at bottom edge
      F(1,j,7) = F(NY,j,7)    ! nw at bottom edge
      F(NY,j,5) = F(1,j,5)    ! south at top edge
      F(NY,j,8) = F(1,j,8)    ! sw at top edge
      F(NY,j,9) = F(1,j,9)    ! se at top edge
    end do
  end subroutine apply_periodic_boundary_conditions
  
  subroutine streaming_step()
    ! Original streaming step: particles move along their velocity directions
    integer :: k, dx, dy
    
    do k = 1, Q
      dx = CX(k)
      dy = CY(k)
      
      ! x shift
      if (dx == 0) then
        TMP2D = F(:,:,k)
      else if (dx == 1) then
        TMP2D(:,2:) = F(:,1:NX-1,k)
        TMP2D(:,1) = F(:,NX,k)
      else if (dx == -1) then
        TMP2D(:,1:NX-1) = F(:,2:,k)
        TMP2D(:,NX) = F(:,1,k)
      end if
      
      ! y shift
      if (dy == 0) then
        F(:,:,k) = TMP2D
      else if (dy == 1) then
        F(2:,:,k) = TMP2D(1:NY-1,:)
        F(1,:,k) = TMP2D(NY,:)
      else if (dy == -1) then
        F(1:NY-1,:,k) = TMP2D(2:,:)
        F(NY,:,k) = TMP2D(1,:)
      end if
    end do
  end subroutine streaming_step
  
  subroutine handle_cylinder_boundary()
    ! Original bounce-back method where particles hitting the cylinder reverse their direction
    integer :: k, i, j
    logical :: incoming_particles(NY,NX)
    
    F2 = F  ! copy distribution functions
    
    do k = 2, Q  ! skip rest particle (k=1)
      ! Find incoming particles: inside cylinder but moving toward cylinder surface
      do j = 1, NX
        do i = 1, NY
          if (CYL(i,j)) then
            ! Check if neighbor in direction -CX(k), -CY(k) is not solid
            incoming_particles(i,j) = .not. CYL(mod(i-CY(k)-1+NY,NY)+1, mod(j-CX(k)-1+NX,NX)+1)
          else
            incoming_particles(i,j) = .false.
          end if
        end do
      end do
      
      ! Bounce back: reverse particle direction
      where (incoming_particles)
        F2(:,:,k) = F(:,:,OPP(k))
      end where
    end do
    
    F = F2  ! update F with bounced-back values
  end subroutine handle_cylinder_boundary
  
  subroutine compute_macroscopic_variables()
    ! Compute macroscopic variables (matching Python)
    RHO = 0.0
    UX = 0.0
    UY = 0.0
    
    do k = 1, Q
      RHO = RHO + F(:,:,k)
      UX = UX + F(:,:,k) * real(CX(k))
      UY = UY + F(:,:,k) * real(CY(k))
    end do
    
    UX = UX / RHO
    UY = UY / RHO
    
    ! Set velocity to zero in cylinder
    where (CYL)
      UX = 0.0
      UY = 0.0
    end where
  end subroutine compute_macroscopic_variables
  
  subroutine collision_step()
    ! Original collision step using the BGK approximation
    real :: Feq(NY,NX,Q)
    integer :: k
    
    ! Compute equilibrium distribution functions
    do k = 1, Q
      cu = real(CX(k))*UX + real(CY(k))*UY
      usq = UX*UX + UY*UY
      Feq(:,:,k) = RHO * W(k) * (1.0 + 3.0*cu + 4.5*cu*cu - 1.5*usq)
    end do
    
    ! BGK collision: relax toward equilibrium
    F = F - (1.0/TAU) * (F - Feq)
  end subroutine collision_step
  
  subroutine apply_inflow_outflow_boundary_conditions()
    ! Apply inlet/outlet boundary conditions (matching Python)
    integer :: i, k
    real :: cu_col(NY)
    
    UX(:,1) = UMAX
    UX(:,NX) = UMAX
    UY(:,1) = 0.0
    UY(:,NX) = 0.0
    
    ! Left edge
    usqL = UX(:,1)*UX(:,1)
    do k = 1, Q
      cu_col = real(CX(k)) * UX(:,1)
      TMP2D(:,1) = 1.0 + 3.0*cu_col + 4.5*cu_col*cu_col - 1.5*usqL
      F(:,1,k) = RHO(:,1) * W(k) * TMP2D(:,1)
    end do
    
    ! Right edge
    usqR = UX(:,NX)*UX(:,NX)
    do k = 1, Q
      cu_col = real(CX(k)) * UX(:,NX)
      TMP2D(:,NX) = 1.0 + 3.0*cu_col + 4.5*cu_col*cu_col - 1.5*usqR
      F(:,NX,k) = RHO(:,NX) * W(k) * TMP2D(:,NX)
    end do
  end subroutine apply_inflow_outflow_boundary_conditions
  
  subroutine plot_vorticity_png(step, picnum)
    integer, intent(in) :: step, picnum
    
    character(len=256) :: fname_ppm
    integer, allocatable :: img(:,:,:)
    integer :: i, j, u
    real :: vort(NY,NX)
    logical :: ex
    
    ! Compute vorticity
    do j = 1, NX
      do i = 1, NY
        vort(i,j) = (UY(i,1 + modulo(j,NX)) - UY(i,1 + modulo(j-2,NX))) - &
                    (UX(1 + modulo(i,NY),j) - UX(1 + modulo(i-2,NY),j))
      end do
    end do
    
    if (picnum == 1) then
      write(*,*) 'vorticity min/max:', minval(vort), maxval(vort)
      write(*,*) 'UX min/max:', minval(UX), maxval(UX)
      write(*,*) 'UY min/max:', minval(UY), maxval(UY)
      write(*,*) 'RHO min/max:', minval(RHO), maxval(RHO)
    end if
    
    ! Allocate image array
    allocate(img(NY,NX,3))
    
    ! Map vorticity to colors (RdBu colormap)
    call map_to_rdBu(vort, -0.02, 0.02, img)
    
    ! Paint cylinder interior black
    do j = 1, NX
      do i = 1, NY
        if (CYL(i,j)) then
          img(i,j,1) = 0
          img(i,j,2) = 0
          img(i,j,3) = 0
        end if
      end do
    end do
    
    ! Draw cylinder outline
    call draw_circle_outline(img, CX0, CY0, R)
    
    ! Ensure output directory exists
    inquire(file='images', exist=ex)
    if (.not. ex) call execute_command_line('mkdir images', wait=.true.)
    
    ! Write PPM file
    write(fname_ppm,'(A,I4.4,A)') 'images/lattice_boltzmann_', picnum, '_fortran.ppm'
    open(newunit=u, file=fname_ppm, status='replace', action='write', form='formatted')
    write(u,'(A)') 'P3'
    write(u,'(I0,1X,I0)') NX, NY
    write(u,'(I0)') 255
    
    ! Write image data (top to bottom for matplotlib compatibility)
    do i = NY, 1, -1
      do j = 1, NX
        write(u,'(I0,1X,I0,1X,I0)') img(i,j,1), img(i,j,2), img(i,j,3)
      end do
    end do
    close(u)
    
    deallocate(img)
  end subroutine plot_vorticity_png
  
  subroutine map_to_rdBu(z, zmin, zmax, img)
    real, intent(in) :: z(:,:), zmin, zmax
    integer, intent(out) :: img(size(z,1), size(z,2), 3)
    integer :: i, j, r, g, b
    real :: t, val
    
    do j = 1, size(z,2)
      do i = 1, size(z,1)
        val = z(i,j)
        t = (val - zmin) / max(1.0e-20, (zmax - zmin))
        t = max(0.0, min(1.0, t))
        call rdBu_color(t, r, g, b)
        img(i,j,1) = r
        img(i,j,2) = g
        img(i,j,3) = b
      end do
    end do
  end subroutine map_to_rdBu
  
  subroutine rdBu_color(t, r, g, b)
    real, intent(in) :: t
    integer, intent(out) :: r, g, b
    real :: r1, g1, b1, r2, g2, b2, u
    
    if (t <= 0.5) then
      u = t / 0.5
      r1 = 49.0;  g1 = 54.0;  b1 = 149.0
      r2 = 255.0; g2 = 255.0; b2 = 255.0
    else
      u = (t - 0.5) / 0.5
      r1 = 255.0; g1 = 255.0; b1 = 255.0
      r2 = 165.0; g2 = 0.0;   b2 = 38.0
    end if
    
    r = int(nint((1.0 - u) * r1 + u * r2))
    g = int(nint((1.0 - u) * g1 + u * g2))
    b = int(nint((1.0 - u) * b1 + u * b2))
    r = max(0, min(255, r))
    g = max(0, min(255, g))
    b = max(0, min(255, b))
  end subroutine rdBu_color
  
  subroutine draw_circle_outline(img, xc, yc, radius)
    integer, intent(inout) :: img(:,:,:)
    integer, intent(in) :: xc, yc, radius
    integer, parameter :: N = 1024
    integer :: k, xi, yi
    real :: th
    
    do k = 0, N-1
      th = 6.283185307179586 * real(k) / real(N)
      xi = nint(real(xc) + real(radius) * cos(th))
      yi = nint(real(yc) + real(radius) * sin(th))
      if (xi >= 1 .and. xi <= size(img,2) .and. yi >= 1 .and. yi <= size(img,1)) then
        img(yi,xi,1) = 0
        img(yi,xi,2) = 0
        img(yi,xi,3) = 0
      end if
    end do
  end subroutine draw_circle_outline

end program lbm_halfway_bounceback